IMPLEMENTATION.md

```diff
@@ -0,0 +1,98 @@
+# 3D Tiles Validator 1.1 Implementation Notes
+
+Parts of the current implementation are still in a **DRAFT** state. 
+
+## Directory structure:
+
+- `./src`: The entry point for the command line interface is in `main.ts`
+- `./src/validation`: The core classes for the 3D Tiles validation
+- `./src/base`: Generic, low-level utility functions
+- `./src/io`: Classes for reading and resolving resources via URIs
+- `./src/issues`: Classes defining the categories of detected issues
+- `./src/structure`: Plain old objects for the 3D Tiles types
+- `./src/tileFormats`: Validators for tile content
+
+- `./src/json`: Classes for generic JSON-schema based validation. This currently only contains an implementation based on `ajv` that is used for internal verification. This is supposed to be extended to support JSON-schema based validation of extensions, with some sort of a "plugin" concept. 
+ 
+- Work in progress: 
+  - `./src/implicitTiling`: Classes that support accessing implicit tiling information
+  - `./src/traversal`: Classes for traversing tilesets
+  - `./src/metadata`: Classes that support accessing 3D Metadata
+
+- `./specs`: Jasmine spec drafts
+- `./specs/data`: Test data
+
+## Overview
+
+The classes in the `./src/validation` directory are the core classes for the 3D Tiles validation. 
+
+The entry point for the tileset validation is the `TilesetValidator` class:
+
+- It receives a tileset JSON string and parses it into a `structure/Tileset.ts` object.
+  - Note: These `structure/*` classes are "plain objects". The do not have methods, and no real type checking. They _only_ hold the parsed data.
+- It traverses the `Tileset` structure, and performs the validation
+  - For each object type, there is a validator, like `AssetValidator`, `ContentValidator`, `TileValidator`, `BoundingVolumeValidator` etc. For now, these are mainly classes with single static functions, but they might be refined for some objects.
+
+Each validation function receives a `ValidationContext` as the last parameter. 
+
+The `ValidationContext` class:
+
+- Can contain configuration options that may be needed in the future (like `doValidateContents=false` or so)
+- Maintains a `ResourceResolver` that reads a `Buffer` from a given URI (resolved in the respective context - the tileset, external tileset, or glTF)
+- Collects the `ValidationIssue` instances (errors and warnings) that are found during the traversal
+
+The `ValidationIssue` class and its types:
+
+- The `ValidationIssue` itself is a plain data structure with basic info about the issue (message, path, severity...)
+- Instances of this class are created with methods in the `issues*` classes, roughly categorized here:
+  - `issues/IoValidationIssues.ts`: Fundamental IO errors, like `JSON_PARSE_ERROR`
```

Comment:
I got this error for one tileset. The `path` field is empty, is that expected? It almost looks like an internal error in the code that manifested as a validation error.

```
(base) slilley@Hydrogen:~/Code/3d-tiles-validator$ npx ts-node src/main.ts --tilesetFile ~/Desktop/CDB_yemen_gsmodel/tileset.json 
Validating tileset /home/slilley/Desktop/CDB_yemen_gsmodel/tileset.json
Validating as external tileset: GSModels_1_1.json
Validating as external tileset: Tiles/N12/E045/GSModels/1_1/N12E045_D300_S001_T001.json
Validating GLB: N12E045_D300_S001_T001_LC5_U0_R0.glb
Validating GLB: N12E045_D300_S001_T001_LC4_U0_R0.glb
Validating GLB: N12E045_D300_S001_T001_LC3_U0_R0.glb
Validating GLB: N12E045_D300_S001_T001_LC2_U0_R0.glb
Validating GLB: N12E045_D300_S001_T001_LC1_U0_R0.glb
Validation result:
{
  "date": "2022-09-21T20:14:46.871Z",
  "numErrors": 1,
  "numWarnings": 0,
  "issues": [
    {
      "type": "JSON_PARSE_ERROR",
      "path": "",
      "message": "TypeError: Cannot read properties of null (reading 'subarray')",
      "severity": "ERROR"
    }
  ]
}
```

Reply:
> I got this error for one tileset. The path field is empty, is that expected? It almost looks like an internal error in the code that manifested as a validation error.

Yes, it is an internal error that is not caught properly. I saw this once, with an implicit tileset where a part of the subtree handling was wrong (roughly: It tried to load a `subtree` file that was already detected to be invalid). 

Two notes:

- The implicit tiling part is not yet finalized, but the more test cases the better. If this is a tileset that is publicly/internally available, it would be good to run an own test. 
- Of course, these errors should appear as a `JSON_PARSE_ERROR`. (That's a too generic `catch(error)` at some point). I'll review these parts, and try to make sure that they only catch actual parsing errors (and report something like an `INTERNAL_ERROR` as a last resort - i.e. as something that should never happen, akin to an "assertion failure")

Reply:
The fix for the _reason_ of the error was added in https://github.com/CesiumGS/3d-tiles-validator/pull/222/commits/caa14ab6856d8e548cbb0283947c2383a0d12f38 , together with some _highly_ preliminary error handling for similar cases in the future. The broader question about the handling of missing/unresolvable resources has been added to the implementation notes.